### PR TITLE
Response map filter

### PR DESCRIPTION
**Pull request by esmet**
HTTP filter for rewriting HTTP responses in the filter chain. Works for both responses originating from an upstream cluster or from another filter deeper in the filter chain (eg, the "router" filter, which may 404 when no route is found).

This PR contains a substantial amount of copy-paste to adapt the existing local reply config user experience to this feature. Long term we'll need to deduplicate the code but for now I believe it's best to copy and modify to get something working more quickly and reliably.

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -87,6 +87,7 @@ extensions/filters/common/original_src @snowp @klarose
 /*/extensions/filters/http/grpc_json_transcoder @qiwzhang @lizan
 /*/extensions/filters/http/router @alyssawilk @mattklein123 @snowp
 /*/extensions/filters/http/ext_authz @gsagula @dio
+/*/extensions/filters/http/error_response @esmet @dio
```
Comment (kflynn) [inline CODEOWNERS]: Did you check with whoever `dio` is about this? 🙂

File: api/envoy/config/core/v3/substitution_format_string.proto [hunk 2]
```diff
@@ -57,5 +60,29 @@ message SubstitutionFormatString {
     //  }
     //
     google.protobuf.Struct json_format = 2 [(validate.rules).message = {required: true}];
+
+    // A DataSource variant for text_format.
+    //
+    // The same semantics apply to the string/ that the referenced DataSource contains.
```
Comment (LukeShu) [inline api/envoy/config/core/v3/substitution_format_string.proto]: > string/

Is that `/` some form of markup that I don't understand?

File: api/envoy/config/core/v3/substitution_format_string.proto [hunk 3]
```diff
@@ -57,5 +60,29 @@ message SubstitutionFormatString {
...
+    // This field takes precedent over text_format_source.
+    DataSource text_format_source = 5;
```
Comment (LukeShu) [inline api/envoy/config/core/v3/substitution_format_string.proto]: Wait, this is `text_format_source`.  Which thing is it taking precedence over?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto]: that is a typo

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 3 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto]: this should say 'over text_format'

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+// Error Response Filter :ref:`configuration overview <config_http_filters_error_response>`.
+// [#extension: envoy.filters.http.error_Response]
+
+// The configuration to filter and change local response.
+message ResponseMapper {
+  // Filter to determine if this mapper should apply.
+  config.accesslog.v3.AccessLogFilter filter = 1 [(validate.rules).message = {required: true}];
+
+  // The new response status code if specified.
+  google.protobuf.UInt32Value status_code = 2 [(validate.rules).uint32 = {lt: 600 gte: 200}];
```
Comment (LukeShu) [inline api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: No 1XX responses?

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  // The new error body text if specified. It will be used in the `%LOCAL_REPLY_BODY%`
+  // command operator in the `body_foramt`.
```
Comment (LukeShu) [inline api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: > foramt

chcek yuor spleling

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: ErrorResponse]
+// Error Response Filter :ref:`configuration overview <config_http_filters_error_response>`.
+// [#extension: envoy.filters.http.error_Response]
+
+// The configuration to filter and change local response.
+message ResponseMapper {
+  // Filter to determine if this mapper should apply.
+  config.accesslog.v3.AccessLogFilter filter = 1 [(validate.rules).message = {required: true}];
```
Comment (LukeShu) [inline api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: Why is this required?

Comment (esmet) [reply api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: The filter represents the "if" portion of "If response matches, then rewrite to ..."

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto (hunk 5 above)
Comment (esmet) [reply api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: looks like there's a typo in api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto's version of this comment, i'll fix it in our copy

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  // .. code-block::
+  //
+  //   text_format: %LOCAL_REPLY_BODY%:%RESPONSE_CODE%:path=$REQ(:path)%
+  //
+  // The following response body in `plain/text` format will be generated for a request with
+  // local reply body of "upstream connection error", response_code=503 and path=/foo.
+  //
+  // .. code-block::
+  //
+  //   upstream connect error:503:path=/foo
```
Comment (LukeShu) [inline api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: "upstream connect[ion] error"?

Comment (esmet) [reply api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: i'll update our copy to include this fix

File: api/envoy/extensions/filters/http/error_response/v3/error_response.proto (hunk 4 above)
Comment (esmet) [reply api/envoy/extensions/filters/http/error_response/v3/error_response.proto]: right, no rewrites on 1xx

File: source/common/response_map/response_map.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+   * @param content_type response content_type.
+   */
+  virtual void rewrite(const Http::RequestHeaderMap* request_headers,
+                       Http::ResponseHeaderMap& response_headers,
+                       StreamInfo::StreamInfo& stream_info, std::string& body,
+                       absl::string_view& content_type) const PURE;
+
+  virtual bool match(const Http::RequestHeaderMap* request_headers,
+                     const Http::ResponseHeaderMap& response_headers,
+                     StreamInfo::StreamInfo& stream_info) const PURE;
```
Comment (kflynn) [inline source/common/response_map/response_map.h:32]: No comments for `match`?

File: source/common/response_map/response_map.cc [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+  BodyFormatter(const envoy::config::core::v3::SubstitutionFormatString& config, Api::Api& api)
+      : formatter_(Formatter::SubstitutionFormatStringUtils::fromProtoConfig(config, api)),
+        content_type_(
+            !config.content_type().empty()
+                ? config.content_type()
+                : config.format_case() ==
+                          envoy::config::core::v3::SubstitutionFormatString::FormatCase::kJsonFormat
+                      ? Http::Headers::get().ContentTypeValues.Json
+                      : Http::Headers::get().ContentTypeValues.Text) {}
```
Comment (kflynn) [inline source/common/response_map/response_map.cc:33]: Wasn't there some discussion of commenting the initializer logic here, since the `Content-Type` stuff is kind of critical?

...oh, huh, wait. Will that screw up merging the stuff you backported this from?

**Review by kflynn — CHANGES_REQUESTED** [review]
I'd like to get the answer on the comment questions, but the code looks fine.

**Review by LukeShu — APPROVED** [review]

File: source/common/response_map/response_map.cc (hunk 9 above)
Comment (esmet) [reply source/common/response_map/response_map.cc:33]: Yeah this is backported

File: source/common/response_map/response_map.h (hunk 8 above)
Comment (esmet) [reply source/common/response_map/response_map.h:32]: done

**Review by kflynn — APPROVED** [review]
Let's go ahead and merge this, and improve the comments as we work with it. It makes me nervous demanding lots of additions given the backporting happening in here.
